README.md
1.1.4

Intro
With this jQuery plugin you can configure what character to use for the numpad decimal separator. This plugin is tested with jsTestDriver and jasmine.

Dependencies
jQuery (http://jquery.com)
jQuery extension functions
numpadDecSeparator this function has three options:
separator: the separator to use when a user presses the numpad decimal separator key. Only used when the option useRegionalSettings is false. Defaults to ','. You can also use one of the predefined variables.
useRegionalSettings: when a user presses the numpad decimal separator key it will use the regional options of the operating system. This only works in firefox and ie 7+!!! Defaults to false.
predefinedVariables - the default predifined variables are:
SPACE => will output space
COMMA => will output comma
numpadDecSeparator('unbind') - this function will unbind the numpadDecSeparator
numpadDecSeparator('version') - static function which returns the current version of the plugin
numpadDecSeparator('mergeDefaults') - static function
with this function you can override some or all the default options, the provided options will be merged with the default options
Examples
$(".amount").numpadDecSeparator();
$(".amount").numpadDecSeparator({separator: ','}); 
//this is the same as 
//$(".amount").numpadDecSeparator({separator: 'COMMA'});

$(".amount").numpadDecSeparator({separator: ' '}); 
//this is the same as 
//$(".amount").numpadDecSeparator({separator: 'SPACE'});

$(".amount").numpadDecSeparator({useRegionalSettings: true});

$(".amount").numpadDecSeparator('unbind');

$.fn.numpadDecSeparator('version');

$.fn.numpadDecSeparator('mergeDefaults', {separator: "SPACE"});
$.fn.numpadDecSeparator('mergeDefaults', {useRegionalSettings: true});
$.fn.numpadDecSeparator('mergeDefaults', {predefinedVariables: {APOSTROPHE: "'"}});
var newDefaults = {
 separator : ' ',
 useRegionalSettings : true,
 predefinedVariables: {SPACE: " "}
};
$.fn.numpadDecSeparator('mergeDefaults', newDefaults);



ChangeLog.md
## Changes in version 1.1.4 (18/12/2011) ##
Refactored tests to BDD jasmine, removed qunit tests.

## Changes in version 1.1.2 (28/04/2011) ##
Added static function mergeDefaults.

Added predifined variables (SPACE and COMMA).

## Changes in version 1.0.1 ##
Minor changes to js code.
Refactored project to use eclipse with:
  * m2eclipse plugin (maven)
  * jsTestDriver plugin
  * Subclipse plugin (svn)

The project has a fully automated maven build.

The build runs on windows 7.

The tests are run in ie, firefox, chrome, opera en safari.

The jsTestDriver tests are run during the build.

The jsTestDriver tests are found in the folder src/test/resources.

The QUnit tests are still located in the tests folder.
They are not run during the build.

Refactored the demo page.

## version 1.0.0 ##
Initial version.
This project is tested with QUnit, these are located in the tests folder of the zip.

src/test/resources/testUtil.js
function triggerNumpadDecimalSeperator(el) {
	var eventKeydown = $.Event("keydown");
	eventKeydown.keyCode = jQuery.browser.opera ? 78 : 110;
	var eventKeypress = $.Event("keypress");
	eventKeypress.charCode = ".".charCodeAt(0);
	el.trigger(eventKeydown).trigger(eventKeypress);
}

src/test/resources/numpadDecSeparatorSpec.js
jasmine.getFixtures().fixturesPath = '/test/test/resources/spec/fixtures';
var version = '1.1.5';
describe('numpadDecSeparator', function () {
	describe('version', function (){
		it('should be correct version', function(){
			expect($.fn.numpadDecSeparator('version')).toEqual(version);
		});
	});
	describe('with default option', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
			$("#testInput").numpadDecSeparator();
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
		});
		
		it('separator should be comma', function() {
			expect($("#testInput").val()).toEqual(',');
		});
	});
	describe('with multiple elements', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
			$("#testInput").numpadDecSeparator();
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
			
			$("#testInput2").numpadDecSeparator({separator: ' '});
			$("#testInput2").focus();
			triggerNumpadDecimalSeperator($("#testInput2"));
		});
		
		it('each should have his own configured separator', function() {
			expect($("#testInput").val()).toEqual(',');
			expect($("#testInput2").val()).toEqual(' ');
		});
	});
	describe('initialize twice', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
			$("#testInput").numpadDecSeparator().numpadDecSeparator();
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
		});
		
		it('should only return one comma', function() {
			expect($("#testInput").val()).toEqual(',');
		});
	});
	describe('replace selection', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
			$("#testInput").numpadDecSeparator();
			$("#testInput").val("12345").setSelection(1,4);
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
		});
		
		it('selected text should be replaced with comma', function() {
			expect($("#testInput").val()).toEqual('1,5');
		});
	});
	describe('readonly element', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
			$("#testInput").attr("readonly", true);
			$("#testInput").numpadDecSeparator({separator: ' '});
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
		});
		
		it('a readonly element should not bind numpadDecSeparator', function() {
			var data = $("#testInput").data('numpadDecSeparator');
			expect(!data).toBeTruthy();
			expect($("#testInput").val()).toEqual('');
		});
	});
	describe('should be able to configure separator', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
		});
		
		it('with a space', function() {
			$("#testInput").numpadDecSeparator({separator: ' '});
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
			
			expect($("#testInput").val()).toEqual(' ');
		});
		
		it('with predifined SPACE variable', function() {
			$("#testInput").numpadDecSeparator({separator: 'SPACE'});
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
			
			expect($("#testInput").val()).toEqual(' ');
		});
		
		it('with predifined COMMA variable', function() {
			$("#testInput").numpadDecSeparator({separator: 'COMMA'});
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
			
			expect($("#testInput").val()).toEqual(',');
		});
		
		it('with an apostrophe', function() {
			$("#testInput").numpadDecSeparator({separator: "\'"});
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
			
			expect($("#testInput").val()).toEqual("'");
		});
	});
	describe('unbind', function (){
		beforeEach(function () {
			loadFixtures('test_fixture.html');
			$("#testInput").numpadDecSeparator({separator: ' '});
			$("#testInput").focus();
			triggerNumpadDecimalSeperator($("#testInput"));
		});
		
		it('should remove data (before unbind data should be present, after unbind data should not be present)', function() {
			expect($("#testInput").val()).toEqual(' ');
			var data = $("#testInput").data('numpadDecSeparator');
			//before unbind data should be present
			expect(!!data).toBeTruthy();
			
			$("#testInput").numpadDecSeparator('unbind');
			
			data = $("#testInput").data('numpadDecSeparator');
			//after unbind data should not be present
			expect(!data).toBeTruthy();
		});
		it('after unbind numpad separator is not space', function() {
			expect($("#testInput").val()).toEqual(' ');
			$("#testInput").numpadDecSeparator('unbind');
			$("#testInput").val("");
			triggerNumpadDecimalSeperator($("#testInput"));
			
			var valueAfterUnbind = $("#testInput").val();
			expect(valueAfterUnbind).not.toEqual(' ');
		});
	});
	describe('mergeDefaults', function (){
		it('should be able to merge separator', function() {
			var pdv = $.fn.numpadDecSeparator.defaults.predefinedVariables;
			
			$.fn.numpadDecSeparator('mergeDefaults', {separator: "SPACE"});
			
			expect($.fn.numpadDecSeparator.defaults.separator).toEqual("SPACE");
			expect($.fn.numpadDecSeparator.defaults.useRegionalSettings).toBeFalsy();
			//merged default predefinedVariables should be the same as before merge
			expect($.fn.numpadDecSeparator.defaults.predefinedVariables).toEqual(pdv);
		});
		it('should be able to merge useRegionalSettings', function() {
			expect($.fn.numpadDecSeparator.defaults.useRegionalSettings).toBeFalsy();
			$.fn.numpadDecSeparator('mergeDefaults', {useRegionalSettings: true});
			expect($.fn.numpadDecSeparator.defaults.useRegionalSettings).toBeTruthy();
		});
		it('should be able to merge predefinedVariables', function() {
			var pdvApos = {APOSTROPHE: "'"};
			expect($.fn.numpadDecSeparator.defaults.predefinedVariables).not.toEqual(pdvApos);
			$.fn.numpadDecSeparator('mergeDefaults', {predefinedVariables: pdvApos});
			expect($.fn.numpadDecSeparator.defaults.predefinedVariables).toEqual(pdvApos);
		});
		it('should be able to merge all', function() {
			var newDefaults = {
					separator : ' ',
					useRegionalSettings : true,
					predefinedVariables: {SPACE: " "}
			};
			$.fn.numpadDecSeparator('mergeDefaults', newDefaults);
			expect($.fn.numpadDecSeparator.defaults).toEqual(newDefaults);
		});
	});
	//this only works in msie and mozilla and regional settings of OS set to nl or equivalent (where decimal seperator is comma)
	if(jQuery.browser.msie || jQuery.browser.mozilla) {
		describe('regional settings', function (){
			it('should be able use regional settings', function() {
				loadFixtures('test_fixture.html');
				$("#testInput").numpadDecSeparator({useRegionalSettings: true});
				$("#testInput").focus();
				triggerNumpadDecimalSeperator($("#testInput"));
				
				//numpad separator should be comma, if test fails => check regional settings of OS => decimal seperator should be comma
				expect($("#testInput").val()).toEqual(',');
			});
			
		});
	}
});





js/jquery.numpadDecSeparator-1.0.0.js
/*
*	Numpad decimal separator plugin for jQuery.
*
*	@author Gert Nuyens
*	@version 1.0.0
*
* 	Dependencies:
*	- jQuery (http://jquery.com)
*
* 	jQuery extension functions:
*	- numpadDecSeparator
*		this function has two options: 
*			1) separator: the separator to use when a user
*				presses the numpad decimal separator key.
*				Only use this when the option useRegionalSettings is false.
*				Defaults to ','.
*			2) useRegionalSettings: when a user presses the numpad decimal 
*				separator key it will use the regional options of the operating system.
*				This only works in firefox and ie!!!
*				Defaults to false.
*	- unNumpadDecSeparator
*		this function will unbind the numpadDecSeparator
*
*	Examples:
*	$(".amount").numpadDecSeparator();
*	$(".amount").numpadDecSeparator({separator: ' '});
*	$(".amount").numpadDecSeparator({useRegionalSettings: true});
*	$(".amount").unNumpadDecSeparator();
*/
(function($){
	$.fn.extend({
		unNumpadDecSeparator: function() { return this.trigger("undo"); },
		numpadDecSeparator: function(options) {
			options = $.extend({
				separator: ',',
				useRegionalSettings: false
			}, options||{});
			return this.each(function() {
				var keydownCode = '';
				var input = $(this);
				if (!input.attr("readonly")){
					input
					.one("undo", function() {
							input
								.unbind(".numpadDecSeparator");
					})
					.bind('keydown.numpadDecSeparator', function(event) {
						keydownCode = event.keyCode;
					})
					.bind('keypress.numpadDecSeparator', function(event) {
						if(_numericPadPeriodPressed(keydownCode) 
							&& !event.shiftKey 
							&& !event.ctrlKey 
							&& !event.altKey) {
							_replaceSelectedVal(this, _getSeparator(options));
							event.preventDefault();     
						}     
					});
				}

			});
		}
	});
	function _decimalSeparator() {
		var n = 1.1;
		n = n.toLocaleString().substring(1, 2);
		return n;
	};
	function _numericPadPeriodPressed(keydownCode){
		if(jQuery.browser.opera) {
			return 78 == keydownCode;
		}
		return 110 == keydownCode;
	};
	function _getSeparator(settings){
		return settings.useRegionalSettings ? _decimalSeparator() : settings.separator;
	};
	function _replaceSelectedVal(input, text){
		if('selectionStart' in input) {
			var start = input.selectionStart+1;
			input.value = input.value.substr(0, input.selectionStart) + text + input.value.substr(input.selectionEnd, input.value.length);
			input.selectionStart = start;
			input.selectionEnd = start;
			input.focus();
		} else if(document.selection) {
			//document.selection.createRange().text = text;
			var sel = document.selection.createRange();
			sel.text = text;
			// Move selection start and end to 0 position
			sel.moveStart ('character', -input.value.length);
			
			// Move selection start and end to desired position
			sel.moveStart ('character', sel.text.length);
			sel.moveEnd ('character', 0);
			sel.select();
		} else {
			input.value += text;
		}
	}
})(jQuery);

src/main/resources/jquery.numpadDecSeparator-min.js
(function(e){var a={init:function(h){return this.each(function(){var i="",k=e(this),j=k.data("numpadDecSeparator");
if(!k.attr("readonly")&&!j){e(this).data("numpadDecSeparator",{target:k});
k.bind("keydown.numpadDecSeparator",function(l){i=l.keyCode
}).bind("keypress.numpadDecSeparator",function(l){if(b(i)&&!l.shiftKey&&!l.ctrlKey&&!l.altKey){f(this,d(h));
l.preventDefault()
}})
}})
},unbind:function(){return this.each(function(){var h=e(this);
h.unbind(".numpadDecSeparator");
h.removeData("numpadDecSeparator")
})
},version:function(){return"1.1.5"
},mergeDefaults:function(h){e.extend(e.fn.numpadDecSeparator.defaults,h)
}};
e.fn.numpadDecSeparator=function(h){var i;
if(a[h]){return a[h].apply(this,Array.prototype.slice.call(arguments,1))
}else{if(typeof h==="object"||!h){if(h){i=e.extend({},e.fn.numpadDecSeparator.defaults,h)
}return a.init.call(this,i?i:e.fn.numpadDecSeparator.defaults)
}else{e.error("Method "+h+" does not exist on jQuery.numpadDecSeparator")
}}};
e.fn.numpadDecSeparator.defaults={separator:",",useRegionalSettings:false,predefinedVariables:{SPACE:" ",COMMA:","}};
function c(){var h=1.1;
h=h.toLocaleString().substring(1,2);
return h
}function b(h){return e.browser.opera?78==h:110==h
}function d(h){return h.useRegionalSettings?c():g(h)
}function g(h){return h.predefinedVariables[h.separator]?h.predefinedVariables[h.separator]:h.separator
}function f(h,j){var k,i;
if("selectionStart" in h){k=h.selectionStart+1;
h.value=h.value.substr(0,h.selectionStart)+j+h.value.substr(h.selectionEnd,h.value.length);
h.selectionStart=k;
h.selectionEnd=k;
h.focus()
}else{if(document.selection){h.focus();
i=document.selection.createRange();
i.text=j;
i.moveStart("character",-h.value.length);
i.moveStart("character",i.text.length);
i.moveEnd("character",0);
i.select()
}else{h.value+=j
}}}})(jQuery);

src/main/resources/jquery.numpadDecSeparator.js
/**
 * Numpad decimal separator plugin for jQuery.
 * 
 * With this jQuery plugin you can configure what character to use for the numpad decimal separator.
 * 
 * @author Gert Nuyens
 * @version 1.1.5 
 * 
 * Dual licensed under the MIT (MIT-LICENSE.txt) 
 * 	or GPL Version 2 licenses (GPL-LICENSE.txt).
 * 
 * Dependencies: - jQuery (http://jquery.com)
 * 
 * jQuery extension functions: 
 * - numpadDecSeparator this function has three options: 
 * 	1) separator: the separator to use when a user presses the numpad decimal separator key. 
 * 		Only use this when the option useRegionalSettings is false. 
 * 		Defaults to ','. You can also use one of the predefined variables.
 * 	2) useRegionalSettings: when a user presses the numpad decimal separator key it will use 
 * 		the regional options of the operating system. 
 * 		This only works in firefox and ie!!! 
 * 		Defaults to false.  
 * 	3) predefinedVariables: the default predifined variables are:
 * 		- SPACE => will output space
 * 		- COMMA => will output comma
 * - numpadDecSeparator('unbind') this function will unbind the numpadDecSeparator
 * - numpadDecSeparator('version') static function which returns the current version of the plugin
 * - numpadDecSeparator('mergeDefaults') static function:
 * 		with this function you can override some or all the default options, the provided options will be
 * 		merged with the default options
 * 
 * Examples: 
 * $(".amount").numpadDecSeparator();
 * $(".amount").numpadDecSeparator({separator: ','}); this is the same as $(".amount").numpadDecSeparator({separator: 'COMMA'});
 * $(".amount").numpadDecSeparator({separator: ' '}); this is the same as $(".amount").numpadDecSeparator({separator: 'SPACE'});
 * $(".amount").numpadDecSeparator({useRegionalSettings: true});
 * 
 * $(".amount").numpadDecSeparator('unbind');
 * 
 * $.fn.numpadDecSeparator('version');
 * 
 * $.fn.numpadDecSeparator('mergeDefaults', {separator: "SPACE"});
 * $.fn.numpadDecSeparator('mergeDefaults', {useRegionalSettings: true});
 * $.fn.numpadDecSeparator('mergeDefaults', {predefinedVariables: {APOSTROPHE: "'"}});
 * var newDefaults = {
 *	separator : ' ',
 *	useRegionalSettings : true,
 *	predefinedVariables: {SPACE: " "}
 * };
 * $.fn.numpadDecSeparator('mergeDefaults', newDefaults);
 **/
(function($) {
	var methods = {
		init : function(options) {
			return this.each(function() {
				var keydownCode = '', $this = $(this), data = $this.data('numpadDecSeparator');
				if (!$this.attr("readonly") && !data) {
					$(this).data('numpadDecSeparator', {
						target : $this
					});
					$this.bind('keydown.numpadDecSeparator', function(event) {
						keydownCode = event.keyCode;
					}).bind(
							'keypress.numpadDecSeparator',
							function(event) {
								if (_numericPadPeriodPressed(keydownCode) && 
										!event.shiftKey && !event.ctrlKey && 
										!event.altKey) {
									_replaceSelectedVal(this,
											_getSeparator(options));
									event.preventDefault();
								}
							});
				}
			});
		},
		unbind : function() {
			return this.each(function() {
				var $this = $(this);
				$this.unbind('.numpadDecSeparator');
				$this.removeData('numpadDecSeparator');
			});
		},
		version : function() {
			return "1.1.5";
		},
		mergeDefaults : function(defaultsToMerge) {
			$.extend($.fn.numpadDecSeparator.defaults, defaultsToMerge);
		}
	};
	$.fn.numpadDecSeparator = function(methodOrOptions) {
		var settings;
		// Method calling logic
		if (methods[methodOrOptions]) {
			return methods[methodOrOptions].apply(this, Array.prototype.slice.call(arguments, 1));
		} else if (typeof methodOrOptions === 'object' || !methodOrOptions) {
			if (methodOrOptions) {
				settings = $.extend({}, $.fn.numpadDecSeparator.defaults, methodOrOptions);
			}
			return methods.init.call(this, settings ? settings : $.fn.numpadDecSeparator.defaults);
		} else {
			$.error('Method ' + methodOrOptions + 
					' does not exist on jQuery.numpadDecSeparator');
		}
	};
	
	$.fn.numpadDecSeparator.defaults = {
		separator : ',',
		useRegionalSettings : false,
		predefinedVariables: {SPACE: " ", COMMA: ","}
	};
	
	function _decimalSeparator() {
		var n = 1.1;
		n = n.toLocaleString().substring(1, 2);
		return n;
	}
	function _numericPadPeriodPressed(keydownCode) {
		return $.browser.opera ? 78 == keydownCode : 110 == keydownCode;
	}
	function _getSeparator(settings) {
		return settings.useRegionalSettings ? _decimalSeparator() : _determineSeparator(settings);
	}
	
	function _determineSeparator(settings){
		return settings.predefinedVariables[settings.separator] ?
				settings.predefinedVariables[settings.separator] : settings.separator;
	}
	function _replaceSelectedVal(input, text) {
		var start, sel;
		if ('selectionStart' in input) {
			start = input.selectionStart + 1;
			input.value = input.value.substr(0, input.selectionStart) + 
							text + 
							input.value.substr(input.selectionEnd, input.value.length);
			input.selectionStart = start;
			input.selectionEnd = start;
			input.focus();
		} else if (document.selection) {
			input.focus();
			sel = document.selection.createRange();
			sel.text = text;
			// Move selection start and end to 0 position
			sel.moveStart('character', -input.value.length);

			// Move selection start and end to desired position
			sel.moveStart('character', sel.text.length);
			sel.moveEnd('character', 0);
			sel.select();
		} else {
			input.value += text;
		}
	}
})(jQuery);